docs/pages/management/admin/labels.mdx

```diff
@@ -280,6 +282,56 @@ computer. Your Teleport user must be authorized to access the server.
    ip-192-168-13-57 ⟵ Tunnel       arch=x86_64,environment=dev,hostname=ip-192-168-13-57
    ```
 
+### Apply resource-based labels
+
+Applying resource-based labels is only supported for servers.
+
+You can apply resource-based labels to a Teleport instance by creating a `server_info`
+resource for the instance. For a server with name `<name>`, its matching `server_info`
+should be named `si-<name>`.
+
+To add resource-based labels:
+
+1. Create the file `server_info.yaml` and paste the following into it:
+
+   ```yaml
+   # server_info.yaml
+   kind: server_info
+   metadata:
+      # Replace "example-node" with the name of the node to apply labels to
+      name: si-example-node
+   spec:
+      new_labels:
+         "foo": "bar"
+   ```
+
+   Run the following to create the `server_info` resource:
+
+   ```code
+   $ tctl create server_info.yaml
+   ```
+
+1. Verify that you have added the label by running the following command on your local
+computer. Your Teleport user must be authorized to access the server. Teleport
+applies labels from `server_info` resources gradually to prevent strain on the Auth Service in
+larger clusters, so it may take several minutes for the new labels to appear.
+
+   ```code
+   $ tsh ls
+   ```
+
+   You should see output similar to the following with the `foo` label displayed:
+
+   ```code
+   Node Name        Address        Labels                                                                                                                   
+   ---------------- -------------- ------------------------------------------------------
+   ip-192-168-13-57 ⟵ Tunnel       foo=bar,hostname=ip-192-168-13-57
+   ```
+
+Resource-based labels are stored in memory on servers, so if a server restarts it may temporarily lose
+its resource-based labels until the Auth Service applies them again. To update resource-based labels,
```

Comment:
> Resource-based labels are stored in memory on servers, so if a server restarts it may temporarily lose
its resource-based labels until the Auth Service applies them again.

This is not the docs question but rather implementation detail which I either missed or forgot about but I'm not sure I understand why this is the case. Aren't these labels supposed to be applied on every heartbeat? If there's a period of time where not all labels are present on the server, this means they effectively can't be used in RBAC's "deny" rules?

Reply:
The auth server has to send the labels to the node over the inventory control stream before the node can report them in future heartbeats. So yes, the shouldn't be used for deny rules.

Reply:
@r0mant This is a pretty tricky feature to track the history of because discussion was spread out all over the place. I think https://github.com/gravitational/teleport/pull/21079#issuecomment-1426231035 is perhaps the most relevant bit, but everything is really spread out.

The tl;dr as I remember it is something like this:
- Loading all ServerInfos immediately either on startup (i.e. cache init) or on first heartbeat would be bad since that had the potentially to effectively double the amount of load experience when restarting a teleport cluster (something that is already problematically intensive and one of our biggest scaling challenges).
- None of the solutions proposed for having nodes refuse connections until all labels had been discovered seemed practical (see linked discussion).
- Its doubtful that there even _is_ a sane way to decide wether or not a node will acquire dynamically applied labels short of requiring that users categorically separate nodes into statically and dynamically labeled groups either via config file options or by using separate join tokens.



> I'm quite a bit worried that with current implementation we just give users ability to shoot themselves in the foot, now you have to track which labels can or can't be used in "deny" rules. 

Oof, I thought we had addressed it, but I don't see it looking back at the code.  I think the fact that we don't have a better story around this is my fault I'm sorry to say.  I was pushing for categorically disallowing dynamic label usage in deny rules early on (https://github.com/gravitational/teleport/pull/21079#issuecomment-1426164237, https://github.com/gravitational/teleport/pull/21079#issuecomment-1426231035), but it looks like that point got lost in the shuffle when that issue was closed.  I was too focused on the scalability discussion in the followup RFD and failed to re-raise the security issue.  Fixing this retroactively without breaking anyone's stuff might be a little tricky 😬

Reply:
Update: I've been giving the deny rule issue more thought.  Here's where my I'm at currently:

Deprecating dynamic labels in deny rules at this point seems like it would be pretty messy.  Its the better fix, but I think that ship has sailed.  Unless anyone has a new idea not covered during the various other issues/discussions related to this, I think label discovery is going to stay lazy, and wether or not a given agent might have dynamically applied labels in the future is likely to continue to be undecidable in most cases.  Given that, my suggestion would be to do one or more of the following:
- Add a warning to role creation/update of the form `Label <label-name> appears to be a cloud label. Use of cloud labels in deny rules is not recommended due to their dynamic nature. Cloud labels are discovered asynchronously, and agents remain accessible even if teleport failed to discover corresponding cloud labels.`
- Add the ability to opt into requiring cloud labels for specific agents statically, either via the agent's own configuration, or via attributes applied to its joining method (e.g. `tctl token add --type=node --require-cloud-labels=yes`).  Basically, have some mechanism of explicitly marking those agents that we know _will_ have cloud labels eventually.  That subset of nodes that joined via discovery could then start pre-configured to require cloud labels, without us breaking nodes that might have cloud labels, but were initially added to the cluster via other means (and therefore might not have cloud labels _known to the auth server_ yet).
- Cache cloud labels on the agent's disk so that nodes can use "last known cloud labels" when restarting, reducing the total amount of time spent without cloud labels (not a fix, just a mitigation).

Reply:
@fspmarshall This is not just about cloud labels, as in https://github.com/gravitational/teleport/pull/34008 we basically added ability for users to set these labels server-side by creating ServerInfo resource.

I don't know if caching these labels on the node helps much tbh because we'll still have this problem when, say, labels are updated while the node is being restarted or down or re-joining with empty data dir.

Instead of (or in addition to) merging the labels during heartbeats, can't we do that at connect time, on-demand? I.e. when connecting to a server, pull its ServerInfo if there's one, and see if there're labels there that aren't on the server? Then it wouldn't put more strain on cache during startup?

Reply:
Ah, didn't realize user-created dynamic labels had landed (jumped into this comment thread without having full context 😬).  That does complicate things.  Especially since, at least from a quick glance, it looks like user-created dynamic labels aren't constrained to a namespace the way cloud labels are so we can't tell which labels might or might not be dynamic from looking at roles/nodes.  Also, as far as I can tell, there isn't any mechanism in-place for cleanup, so user-created labels apply until the node restarts even if the `ServerInfo` is deleted, which is its own kind of problem.

Loading server infos at the time when an access-control decision is needed is definitely appealing in terms of mitigating load on restart, though an additional round trip to auth for every access-control decision wouldn't be ideal.  We have a lot of users requesting that we do whatever we can to _reduce_ resource access latency.  Also, if the point of doing that would be to make using deny rules with dynamic labels sane, then we'd really need to also apply dynamic labels to all heartbeats, and that means the benefits versus traditional caching might be minimal (it might even be worse, since presumably most backends are more performant per-item read when serving range requests versus individual reads, hard to say for sure without testing).

I've been running through various scenarios for the past couple hours, and I don't see any good ways to have our cake and eat it too here. I'll keep mulling it over, but right now I'm thinking that we're going to end up needing to pick a direction and commit to it:

1. Lazy dynamic labels that shouldn't be used for deny rules (likely requiring some trickery to reliably detect and warn about use of dynamic labels in deny rules now that user-created labels have landed and they are un-namespaced).

2. Always consistent labels that significantly increase peak load when in general use (with the risk of serious production outages if anyone currently making liberal use of dynamic labels is also pushing their backend to its limit).

I lean toward option 1.  Not only for itself, but because I find the precedence set by option 2 scary.  There's plenty of room in teleport for more lazy features that can be safely rate-limited without breaking their contracts.  There is very little room for sensitive cross-resource synchronization in the "hot path" so to speak.